<a href="https://colab.research.google.com/github/whosoevernotme/vibecode101/blob/main/lab02_vibe_coding.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/wdittaya/AI-for-digital-health-course/blob/main/lab02_vibe_coding.ipynb)

# Lab 2: Vibe Coding ในงานวิจัยด้านปัญญาประดิษฐ์และการเรียนรู้ของเครื่อง (AI/ML)

**รายวิชา:** 3099704 AI for Digital Health (2026) · **คาบเรียน:** วันศุกร์ที่ 2 ตุลาคม 2026 · **เวลาโดยประมาณ:** 60–90 นาที

## วัตถุประสงค์การเรียนรู้

- ใช้ผู้ช่วยเขียนโค้ดปัญญาประดิษฐ์ (AI coding assistant) สร้างไปป์ไลน์ (pipeline) การเรียนรู้ของเครื่องที่สมบูรณ์จากคำสั่งภาษาธรรมชาติ (prompt)
- ตรวจสอบ (audit) ความถูกต้องของโค้ดที่ปัญญาประดิษฐ์สร้างขึ้น ได้แก่ การรั่วไหลของข้อมูล (data leakage) การเลือกตัวชี้วัดไม่เหมาะสม และข้อผิดพลาดที่ไม่ปรากฏเป็นข้อความแจ้งเตือน (silent bug)
- พัฒนาขั้นตอนการทำงานของตนเอง: ออกคำสั่ง (prompt) → รัน → ตรวจสอบ → ปรับปรุง
- เข้าใจข้อกำหนดด้านความสามารถในการทำซ้ำ (reproducibility) ของงานวิจัยที่ใช้ปัญญาประดิษฐ์ช่วย

> **แนวทางการทำปฏิบัติการ** (งานเดี่ยว)
> - ✅ **ส่วนหลัก — สำหรับผู้เรียนทุกคน ไม่จำเป็นต้องมีพื้นฐานการเขียนโปรแกรม** ให้รันเซลล์ (cell) ตามลำดับจากบนลงล่าง (คลิกปุ่ม ▶ หรือกด **Shift+Enter**) ในปฏิบัติการนี้ *ผู้ช่วยปัญญาประดิษฐ์เป็นผู้เขียนโค้ด* ผู้เรียนเพียงคัดลอกคำสั่งที่เตรียมไว้ รันโค้ดที่ได้รับ และตรวจสอบด้วยรายการคำถามแบบ ใช่/ไม่ใช่ จากผลลัพธ์ที่แสดง แบบฝึกหัด ⭐ แต่ละข้อมีเซลล์ ✅ **เฉลย** อยู่ถัดไป ผู้เรียนควรทำแบบฝึกหัดด้วยตนเองก่อน หากไม่สามารถดำเนินการต่อได้ (หรือโค้ดจากปัญญาประดิษฐ์รันไม่สำเร็จ) ให้รันเซลล์เฉลยแล้วดำเนินการต่อ คำตอบในเซลล์ข้อความ ✍️ คือชิ้นงานที่ต้องส่ง
> - 🏆 **โจทย์ท้าทายบนตารางอันดับ — ไม่บังคับ สำหรับผู้เรียนที่เขียนโปรแกรมได้** อยู่ท้ายโน้ตบุ๊ก (notebook) ให้คะแนนอัตโนมัติบน[ตารางอันดับ (leaderboard) ของรายวิชาบน Hugging Face](https://huggingface.co/spaces/wdittaya/aidh-leaderboard) และไม่นับเป็นส่วนหนึ่งของชิ้นงานที่ต้องส่ง
>
> ทุกขั้นตอนรันได้บน [Google Colab](https://colab.research.google.com) รุ่นไม่มีค่าใช้จ่าย โดยใช้ CPU เพียงอย่างเดียว (ไม่จำเป็นต้องใช้ GPU) หากพบปัญหา สามารถสอบถามผู้สอนผ่านช่องแชตของ Zoom

> 🔒 **ข้อกำหนดด้านความเป็นส่วนตัว — โปรดอ่านก่อนเริ่ม** ห้ามนำข้อมูลผู้ป่วยจริง (ชื่อ HN บันทึกทางคลินิก ผลการตรวจทางห้องปฏิบัติการ หรือไฟล์ข้อมูลจากหน่วยงาน) ไปวางใน Gemini, ChatGPT หรือผู้ช่วยปัญญาประดิษฐ์ใด ๆ ให้ *อธิบาย* ลักษณะข้อมูลแทน (ชื่อคอลัมน์ ชนิดข้อมูล ความหมายของ label) โดยไม่วางข้อมูลรายแถว ปฏิบัติการนี้ใช้ชุดข้อมูล (dataset) โรคเบาหวาน Pima ซึ่งเป็นข้อมูลสาธารณะที่ไม่สามารถระบุตัวบุคคลได้

In [ ]:
# @title ⚙️ การตั้งค่า — โปรดรันเซลล์นี้ก่อน (NAME และ JOIN_CODE ใช้เฉพาะโจทย์ท้าทายบนตารางอันดับ ซึ่งไม่บังคับ)
!pip -q install gradio_client huggingface_hub

HF_OWNER  = "wdittaya"     # บัญชี Hugging Face ที่ใช้โฮสต์ตารางอันดับของรายวิชา (ผู้สอนเป็นผู้กำหนด)
LAB_ID    = "lab02"
NAME      = "whosoever"            # ไม่บังคับ: ชื่อที่จะแสดงบนตารางอันดับแบบสาธารณะ (สามารถใช้นามแฝงได้)
JOIN_CODE = "3099704-dwn"            # ไม่บังคับ: รหัสเข้าร่วมซึ่งผู้สอนประกาศในการเรียนการสอนผ่าน Zoom

SPACE_ID  = f"{HF_OWNER}/aidh-leaderboard"
DATA_REPO = f"{HF_OWNER}/aidh-lab-data"

from huggingface_hub import hf_hub_download

def lab_data(path):
    """ดาวน์โหลดไฟล์ข้อมูลทดสอบสาธารณะของรายวิชา 1 ไฟล์ (เช่น 'pima/test_ids.csv') และคืนค่าตำแหน่งไฟล์ในเครื่อง"""
    return hf_hub_download(DATA_REPO, path, repo_type="dataset")

def submit_to_leaderboard(path, lab=LAB_ID):
    """ส่งไฟล์ผลการทำนายไปยังตารางอันดับของรายวิชา และแสดงคะแนนพร้อมอันดับ"""
    from gradio_client import Client, handle_file
    assert NAME and JOIN_CODE, "โปรดระบุ NAME และ JOIN_CODE ในเซลล์การตั้งค่าก่อน (จำเป็นเฉพาะการส่งผลไปยังตารางอันดับ)"
    client = Client(SPACE_ID, verbose=False)
    print(client.predict(lab=lab, name=NAME, code=JOIN_CODE, file=handle_file(path), api_name="/submit"))

print(f"Leaderboard: https://huggingface.co/spaces/{SPACE_ID}  |  lab={LAB_ID}  name={NAME or '(ยังไม่ได้ระบุ — ตารางอันดับไม่บังคับ)'}")

## Vibe coding คืออะไร

Vibe coding คือการอธิบายสิ่งที่ต้องการเป็นภาษาธรรมชาติ ให้ผู้ช่วยปัญญาประดิษฐ์เขียนโค้ด แล้วปรับแก้ซ้ำหลายรอบ
แนวทางนี้เอื้อต่อการสร้างต้นแบบอย่างรวดเร็ว และทำให้ผู้ที่ไม่มีพื้นฐานการเขียนโปรแกรมสามารถสร้างโมเดลที่ใช้งานได้ แต่ก่อให้เกิดความเสี่ยงหากผู้ใช้ไม่สามารถตรวจสอบผลลัพธ์ได้
ในปฏิบัติการนี้ ผู้เรียนจะให้ปัญญาประดิษฐ์สร้างไปป์ไลน์ จากนั้นตรวจสอบและเปรียบเทียบกับไปป์ไลน์อ้างอิงที่ถูกต้อง

**ผู้ช่วยปัญญาประดิษฐ์ที่ใช้ได้โดยไม่มีค่าใช้จ่าย (เลือกหนึ่งรายการ):**
- **Gemini in Colab** — มีอยู่ใน Google Colab ให้คลิก **ไอคอน Gemini ✨** (มุมขวาบนของ Colab) เพื่อเปิดแผงสนทนา หรือในเซลล์โค้ดว่าง ให้คลิกคำว่า **"generate"** ในข้อความสีเทา *"Start coding or generate with AI"*
- [Google AI Studio](https://aistudio.google.com/)
- ChatGPT / Claude / Copilot รุ่นไม่มีค่าใช้จ่ายบนเว็บ

**หลักปฏิบัติสำคัญ**
1. ไม่เชื่อถือโค้ดที่ยังไม่ได้ตรวจสอบ
2. ไม่เชื่อถือตัวเลขที่อธิบายที่มาไม่ได้
3. บันทึกทั้งคำสั่ง (prompt) และโค้ดไว้ เพื่อเป็นบันทึกการทดลองที่ทำซ้ำได้ (reproducibility)
4. ห้ามนำข้อมูลผู้ป่วยไปวางในผู้ช่วยปัญญาประดิษฐ์ (ดูข้อกำหนดด้านความเป็นส่วนตัวข้างต้น)

In [1]:
# โหลดข้อมูล — ชุดข้อมูลโรคเบาหวาน Pima ชุดเดียวกับ Lab 1 (ผู้หญิง 768 ราย, ข้อมูลสาธารณะ)
import pandas as pd, numpy as np

URL = "https://raw.githubusercontent.com/plotly/datasets/master/diabetes.csv"
df = pd.read_csv(URL)
print(df.shape)
df.head(3)

(768, 9)


,Pregnancies,Glucose,BloodPressure,SkinThickness,Insulin,BMI,DiabetesPedigreeFunction,Age,Outcome
0,6,148,72,35,0,33.6,0.627,50,1
1,1,85,66,29,0,26.6,0.351,31,0
2,8,183,64,0,0,23.3,0.672,32,1


💡 **คำอธิบายผลลัพธ์** ตาราง `df` ประกอบด้วยผู้ป่วยหนึ่งรายต่อหนึ่งแถว มีค่าที่วัดได้ 8 ค่า (glucose, BMI, อายุ ฯลฯ) และคอลัมน์ `Outcome` (1 = เป็นโรคเบาหวาน, 0 = ไม่เป็นโรคเบาหวาน) ผู้ช่วยปัญญาประดิษฐ์ต้องการเพียง *คำอธิบาย* ข้อมูลนี้ โดยไม่จำเป็นต้องเห็นข้อมูลรายแถว

## ขั้นที่ 1 — สร้างโค้ด (Generate)

1. เปิดผู้ช่วยปัญญาประดิษฐ์ (เช่น แผง Gemini ✨ ใน Colab)
2. คัดลอกคำสั่งด้านล่าง **ตามต้นฉบับทุกตัวอักษร** แล้วส่ง

```text
I am working in Google Colab. A pandas DataFrame called `df` is already loaded with the
Pima Indians Diabetes dataset. Columns: Pregnancies, Glucose, BloodPressure, SkinThickness,
Insulin, BMI, DiabetesPedigreeFunction, Age, and the binary label Outcome (1 = diabetes).
Write ONE Python code cell using scikit-learn that: replaces impossible zeros with NaN and
imputes them, splits into train/test, trains a random forest, and reports performance
suitable for a medical screening task. Use the existing `df`; do not download any data and
do not overwrite `df`.
```

   *(คำสั่งนี้เขียนเป็นภาษาอังกฤษเพื่อให้ผู้เรียนทุกคนได้ผลลัพธ์ใกล้เคียงกันและสอดคล้องกับชื่อคอลัมน์ของข้อมูล ในการใช้งานจริง ผู้เรียนสามารถเขียนคำสั่งเป็นภาษาไทยได้)*

3. คัดลอกโค้ดที่ได้รับไปวางในเซลล์ **แบบฝึกหัดที่ 1 (Exercise 1)** ด้านล่าง (แทนที่ `...`) **โดยไม่แก้ไข** แล้วรันเซลล์
4. หากเกิดข้อผิดพลาด (error) ให้ส่งข้อความแจ้งข้อผิดพลาดกลับไปยังผู้ช่วยปัญญาประดิษฐ์ (เช่น *"I got this error: … please fix it"* หรือ *"รันแล้วพบข้อผิดพลาดดังนี้: … โปรดแก้ไข"*) แล้วทดลองอีกหนึ่งครั้ง หากยังไม่สำเร็จ ให้รันเซลล์ ✅ เฉลย ด้านล่างแทน ซึ่งเป็นตัวอย่าง *คำตอบที่ผู้ช่วยปัญญาประดิษฐ์มักให้* เพื่อใช้ในการตรวจสอบขั้นต่อไป

> เหตุที่ใช้คำสั่งนี้: เป็นรูปแบบที่ผู้ใช้ส่วนใหญ่มักใช้ คือสั้นและดูสมเหตุสมผล แต่ *จงใจให้ไม่สมบูรณ์* ในขั้นที่ 2 ผู้เรียนจะพบว่าคำสั่งนี้ขาดข้อกำหนดใดบ้าง

In [ ]:
# ⭐ Exercise 1 — แบบฝึกหัดที่ 1: วางโค้ดที่ปัญญาประดิษฐ์สร้างไว้ด้านล่างโดยไม่แก้ไข แล้วรันเซลล์นี้
# (บันทึกคำสั่งที่ใช้จริงไว้ในคอมเมนต์เพื่อให้ทำซ้ำได้ หากมีการเปลี่ยนคำสั่ง ให้แก้ไขข้อความในส่วนนี้ด้วย)

# PROMPT ที่ใช้: "I am working in Google Colab. A pandas DataFrame called `df` ... (ดูขั้นที่ 1)"
# ผู้ช่วยปัญญาประดิษฐ์ที่ใช้: .Gemini in Colab, วันที่ 03/10/2026..   (เช่น Gemini in Colab, วันที่)

# โค้ดที่ปัญญาประดิษฐ์สร้าง (วางใต้บรรทัดนี้):
...

In [3]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.impute import SimpleImputer
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix, roc_auc_score

# 1. Copy data to avoid overwriting original df
data = df.copy()

# 2. Replace impossible zeros with NaN for specified columns
impossible_zero_cols = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
data[impossible_zero_cols] = data[impossible_zero_cols].replace(0, np.nan)

# 3. Impute missing values with Median
imputer = SimpleImputer(strategy="median")
data[impossible_zero_cols] = imputer.fit_transform(data[impossible_zero_cols])

# 4. Split into features and target
X_ai = data.drop(columns=["Outcome"])
y_ai = data["Outcome"]
X_train, X_test, y_train, y_test = train_test_split(X_ai, y_ai, test_size=0.2, random_state=42)

# 5. Train Random Forest Classifier
model = RandomForestClassifier(n_estimators=100, random_state=42)
model.fit(X_train, y_train)

# 6. Evaluate and report performance suitable for medical screening
y_pred = model.predict(X_test)
y_prob = model.predict_proba(X_test)[:, 1]

tn, fp, fn, tp = confusion_matrix(y_test, y_pred).ravel()
sensitivity = tp / (tp + fn)
specificity = tn / (tn + fp)

print("=== Performance Metrics ===")
print(f"Accuracy: {accuracy_score(y_test, y_pred):.3f}")
print(f"ROC AUC: {roc_auc_score(y_test, y_prob):.3f}")
print(f"Sensitivity (Recall for Diabetes): {sensitivity:.3f}")
print(f"Specificity: {specificity:.3f}")
print("\n--- Confusion Matrix ---")
print(f"True Negative (TN): {tn}  |  False Positive (FP): {fp}")
print(f"False Negative (FN): {fn}  |  True Positive (TP): {tp}")
print("\n--- Classification Report ---")
print(classification_report(y_test, y_pred))

=== Performance Metrics ===
Accuracy: 0.747
ROC AUC: 0.834
Sensitivity (Recall for Diabetes): 0.673
Specificity: 0.788

--- Confusion Matrix ---
True Negative (TN): 78  |  False Positive (FP): 21
False Negative (FN): 18  |  True Positive (TP): 37

--- Classification Report ---
              precision    recall  f1-score   support

           0       0.81      0.79      0.80        99
           1       0.64      0.67      0.65        55

    accuracy                           0.75       154
   macro avg       0.73      0.73      0.73       154
weighted avg       0.75      0.75      0.75       154



In [ ]:
# @title ✅ เฉลย — หากไม่สามารถทำแบบฝึกหัดได้ ให้รันเซลล์นี้ (ดับเบิลคลิกเพื่อแสดงโค้ด) { display-mode: "form" }
# ตัวอย่างคำตอบที่ผู้ช่วยปัญญาประดิษฐ์มักให้สำหรับคำสั่งในขั้นที่ 1 (รันได้แต่ไม่สมบูรณ์ โดยเจตนา
# เพื่อใช้ตรวจสอบในขั้นที่ 2 ในกรณีที่โค้ดของผู้เรียนรันไม่สำเร็จ)
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

data = df.copy()

# Replace impossible zeros with NaN
cols_with_zeros = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
data[cols_with_zeros] = data[cols_with_zeros].replace(0, np.nan)

# Impute missing values with the median
data[cols_with_zeros] = data[cols_with_zeros].fillna(data[cols_with_zeros].median())

# Split into features and target
X_ai = data.drop(columns=["Outcome"])
y_ai = data["Outcome"]
X_train, X_test, y_train, y_test = train_test_split(X_ai, y_ai, test_size=0.2)

# Train a random forest
model = RandomForestClassifier(n_estimators=100)
model.fit(X_train, y_train)

# Evaluate
y_pred = model.predict(X_test)
print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion matrix:\n", confusion_matrix(y_test, y_pred))
print("\nClassification report:\n", classification_report(y_test, y_pred))

💡 **คำอธิบายผลลัพธ์** ผู้ช่วยปัญญาประดิษฐ์สร้างไปป์ไลน์ครบทุกขั้นตอนภายในเวลาไม่กี่วินาที ได้แก่ แก้ไขค่าศูนย์ที่เป็นไปไม่ได้ → เติมค่าที่ขาดหาย (imputation) → แบ่งผู้ป่วยเป็นชุดฝึก (training set) สำหรับให้โมเดลเรียนรู้ และชุดทดสอบ (test set) ที่แยกไว้สำหรับประเมินโมเดล → ฝึกโมเดล random forest (ต้นไม้ตัดสินใจหลายต้นที่ลงมติร่วมกัน) → แสดงคะแนน ขั้นตอนถัดไปคือการตรวจสอบว่าผลลัพธ์นี้ **ถูกต้อง** หรือไม่

## ขั้นที่ 2 — ตรวจสอบ (Audit) ด้วยรายการคำถามแบบ ใช่/ไม่ใช่

โค้ดการเรียนรู้ของเครื่องที่ปัญญาประดิษฐ์สร้างขึ้นมักรันได้โดยไม่เกิดข้อผิดพลาด แต่มีความผิดพลาดแฝงที่ทำให้ผลดูดีกว่าความเป็นจริง ผู้เรียน **ไม่จำเป็น** ต้องแก้ไขโค้ด เพียงพิจารณาผลลัพธ์ของเซลล์แบบฝึกหัดที่ 1 (และค้นหาคำบางคำในโค้ด) แล้วตอบว่า ใช่/ไม่ใช่

ขั้นแรก ให้รันเซลล์ช่วยตรวจสอบด้านล่าง ซึ่งจะแสดงตัวเลขสองค่าที่ใช้ในรายการตรวจสอบ

In [6]:
# เซลล์ช่วยตรวจสอบ — แสดงข้อเท็จจริงที่ใช้ในรายการตรวจสอบ (รันเซลล์นี้เท่านั้น)
share = df["Outcome"].value_counts(normalize=True)
print(f"ผู้ป่วยที่ไม่เป็นโรคเบาหวาน: {share[0]:.1%}   |   ผู้ป่วยที่เป็นโรคเบาหวาน: {share[1]:.1%}")
print(f"=> 'โมเดล' ที่ทำนายว่า 'ไม่เป็นโรคเบาหวาน' สำหรับทุกราย มีค่า accuracy {share[0]:.2f}")
print(f"   แต่ไม่สามารถตรวจพบผู้ป่วยโรคเบาหวานแม้แต่รายเดียว (sensitivity = 0)")
print()
print("ตำแหน่งของการเติมค่าที่ขาดหายและการแบ่งข้อมูล (สำหรับข้อ E)")
print("  ในโค้ดแบบฝึกหัดที่ 1 ให้หาบรรทัดที่มี  train_test_split  และบรรทัดที่มี")
print("  fillna / SimpleImputer / fit_transform / StandardScaler")

ผู้ป่วยที่ไม่เป็นโรคเบาหวาน: 65.1%   |   ผู้ป่วยที่เป็นโรคเบาหวาน: 34.9%
=> 'โมเดล' ที่ทำนายว่า 'ไม่เป็นโรคเบาหวาน' สำหรับทุกราย มีค่า accuracy 0.65
   แต่ไม่สามารถตรวจพบผู้ป่วยโรคเบาหวานแม้แต่รายเดียว (sensitivity = 0)

ตำแหน่งของการเติมค่าที่ขาดหายและการแบ่งข้อมูล (สำหรับข้อ E)
  ในโค้ดแบบฝึกหัดที่ 1 ให้หาบรรทัดที่มี  train_test_split  และบรรทัดที่มี
  fillna / SimpleImputer / fit_transform / StandardScaler


💡 **คำอธิบายผลลัพธ์** ผู้ป่วยประมาณร้อยละ 65 ไม่เป็นโรคเบาหวาน ดังนั้น **การพิจารณาเฉพาะค่า accuracy อาจทำให้เข้าใจผิด** โมเดลอาจมีค่า accuracy ร้อยละ 65 โดยไม่สามารถตรวจพบผู้ป่วยโรคเบาหวานได้แม้แต่รายเดียว สำหรับการตรวจคัดกรอง (screening) ตัวชี้วัดที่สำคัญคือ **ความไว (sensitivity)** หมายถึงสัดส่วนของผู้ป่วยโรคเบาหวานที่โมเดลตรวจพบ และ **ความจำเพาะ (specificity)** หมายถึงสัดส่วนของผู้ที่ไม่เป็นโรคซึ่งโมเดลจำแนกได้ถูกต้อง ส่วน **ROC AUC** แสดงความสามารถของโมเดลในการ *จัดลำดับ* ความเสี่ยงของผู้ป่วย (0.5 = เทียบเท่าการสุ่ม, 1.0 = สมบูรณ์)

### ✅ รายการตรวจสอบ (Audit checklist)

ให้พิจารณาผลลัพธ์ (และโค้ด) ของเซลล์แบบฝึกหัดที่ 1 แล้วตอบแต่ละข้อว่า **ใช่ / ไม่ใช่**

| # | คำถาม | ตำแหน่งที่ต้องพิจารณา | 🚩 สัญญาณเตือนเมื่อตอบ | ความสำคัญในงานด้านสุขภาพ |
|---|---|---|---|---|
| A | มีการรายงาน **ความไว (sensitivity)** หรือไม่ (= *recall* ของคลาส `1`) | ผลลัพธ์: ค่า `recall` ในแถวของ `1` หรือคำว่า *sensitivity* | **ไม่ใช่** | การตรวจคัดกรองต้องไม่พลาดผู้ป่วย (ผลลบลวง, false negative) |
| B | มีการรายงาน **ความจำเพาะ (specificity)** หรือไม่ (หรือมี confusion matrix ที่ใช้คำนวณได้) | ผลลัพธ์ | **ไม่ใช่** | ผลบวกลวง (false positive) จำนวนมากเพิ่มภาระแก่หน่วยบริการ |
| C | มีการรายงาน **ROC AUC** หรือไม่ | ผลลัพธ์ | **ไม่ใช่** | accuracy และ recall ขึ้นกับเกณฑ์ตัดสิน (threshold) ค่าเดียว ขณะที่ AUC ไม่ขึ้นกับเกณฑ์ |
| D | ค่า **accuracy** ที่รายงานสูงกว่าค่าของการทำนาย "ไม่เป็นโรคทุกราย" ที่เซลล์ช่วยตรวจสอบแสดงไว้ (~0.65) อย่างชัดเจนหรือไม่ | ผลลัพธ์เปรียบเทียบกับเซลล์ช่วยตรวจสอบ | **ไม่ใช่** | แสดงว่าโมเดลแทบไม่ดีกว่าการไม่ใช้โมเดล |
| E | มีการเติมค่าที่ขาดหาย (`fillna`, `SimpleImputer`, `fit_transform`, scaler) กับ **ข้อมูลทั้งหมด ในบรรทัดที่อยู่ก่อน** บรรทัด `train_test_split` หรือไม่ | โค้ด — เปรียบเทียบตำแหน่งบรรทัด | **ใช่** | *การรั่วไหลของข้อมูล (data leakage)*: ข้อมูลของชุดทดสอบส่งผลต่อการฝึก ผลจึงดีกว่าความเป็นจริง |
| F | มี `random_state=` ทั้งใน `train_test_split(...)` **และ** ในโมเดลหรือไม่ | โค้ด — ค้นหาคำนี้ | **ไม่ใช่** | ทำซ้ำไม่ได้: ผู้อื่นที่รันโค้ดเดียวกันจะได้ตัวเลขต่างกัน |
| G | เมื่อรันเซลล์แบบฝึกหัดที่ 1 ซ้ำ ตัวเลข **เปลี่ยนไป** หรือไม่ | ผลลัพธ์จากการรันสองครั้ง | **ใช่** | เช่นเดียวกับข้อ F ผลที่ทำซ้ำไม่ได้ไม่สามารถนำไปตีพิมพ์ได้ |
| H | ผลลัพธ์มาจากการแบ่งข้อมูลแบบสุ่ม **ครั้งเดียว** โดยไม่มีการตรวจสอบไขว้ (cross-validation) หรือไม่ | โค้ด — ค้นหาคำว่า `cross_val` | **ใช่** | ชุดทดสอบมีผู้ป่วยประมาณ 150 ราย การแบ่งข้อมูลเพียงครั้งเดียวอาจให้ผลที่คลาดเคลื่อน |

> ไม่บังคับ: ผู้เรียนสามารถส่งคำสั่งถึงผู้ช่วยปัญญาประดิษฐ์ว่า *"โปรดอธิบายโค้ดนี้ทีละบรรทัดสำหรับพยาบาลที่ไม่มีพื้นฐานการเขียนโปรแกรม และระบุว่ามีการ fit imputer ก่อนการแบ่ง train/test หรือไม่"* อย่างไรก็ตาม **ต้องตรวจสอบคำตอบกับรายการตรวจสอบนี้ทุกครั้ง** เนื่องจากผู้ช่วยปัญญาประดิษฐ์มักยืนยันความถูกต้องของโค้ดที่ตนสร้าง

✍️ โปรดบันทึกคำตอบของรายการตรวจสอบ (และตัวเลขสำคัญที่เซลล์แบบฝึกหัดที่ 1 แสดง)

✍️ **คำตอบ:**

- ผู้ช่วยปัญญาประดิษฐ์ที่ใช้ / โค้ดรันสำเร็จในครั้งแรกหรือไม่: Gemini  / ไม่สำเร็จตั้งแต่ครั้งแรก
- A (รายงาน sensitivity): 67.3%
- B (รายงาน specificity): 78.8%
- C (รายงาน ROC AUC): 0.834
- D (accuracy สูงกว่า ~0.65 อย่างชัดเจน): ใช่ เพราะ Accuracy = 0.747 หรือ 74.7% สูงกว่า baseline ประมาณ 65% ราว 9.7 percentage points
- E (เติมค่าก่อนแบ่งข้อมูล — data leakage): มี Data Leakage
- F (มี random_state ครบถ้วน):มี
- G (ตัวเลขเปลี่ยนเมื่อรันซ้ำ):โดยปกติไม่เปลี่ยน
- H (แบ่งข้อมูลเพียงครั้งเดียว): ใช่
- ตัวเลขที่แสดง (accuracy, recall ของคลาส 1 ฯลฯ): Accuracy = 0.747, Recall class 1 = 0.673, Precision class 1 = 0.64, F1-score class 1 = 0.65, ROC AUC = 0.834, Specificity = 0.788

## ขั้นที่ 3 — แก้ไข: เปรียบเทียบกับไปป์ไลน์อ้างอิง

ด้านล่างเป็นการเขียนไปป์ไลน์เดียวกันโดย **ไม่มีการรั่วไหลของข้อมูล** หลักสำคัญคือ `Pipeline` ของ scikit-learn ซึ่งรวมขั้นตอน "เติมค่าที่ขาดหาย → ฝึกโมเดล" ไว้ในออบเจกต์เดียว ค่ามัธยฐาน (median) ที่ใช้เติมจึงคำนวณจาก **ผู้ป่วยในชุดฝึกเท่านั้น** ทุกครั้ง รวมถึงภายในการตรวจสอบไขว้

*การตรวจสอบไขว้ (cross-validation) แบบ 5-fold* คือการแบ่งข้อมูลเป็น 5 ส่วน ใช้ 4 ส่วนในการฝึกและอีก 1 ส่วนในการทดสอบ ทำซ้ำ 5 รอบแล้วหาค่าเฉลี่ย ทำให้ได้คะแนนที่มีเสถียรภาพมากกว่าการแบ่งข้อมูลครั้งเดียว

**⭐ แบบฝึกหัดที่ 2 (Exercise 2) — เติมช่องว่าง 5 ช่อง** (แต่ละช่องเป็นคำเดียวหรือสูตรสั้น คำใบ้ระบุสิ่งที่ต้องพิมพ์ไว้แล้ว) หากไม่สามารถทำได้ ให้รันเซลล์ ✅ เฉลย ซึ่งเป็นไปป์ไลน์อ้างอิงฉบับสมบูรณ์

In [19]:
# ⭐ Exercise 2 — แบบฝึกหัดที่ 2: ไปป์ไลน์อ้างอิงที่ไม่มี data leakage — เติมช่องว่าง 5 ช่องที่เป็น ...
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.metrics import confusion_matrix

ZERO_MEANS_MISSING = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
X = df.drop(columns=["Outcome"]).replace({c: {0: np.nan} for c in ZERO_MEANS_MISSING})
y = df["Outcome"]

pipe = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),                   # ช่องที่ 1: พิมพ์ "median" (รวมเครื่องหมายอัญประกาศ)
    ("model", RandomForestClassifier(n_estimators=300, random_state=42)),   # ช่องที่ 2: พิมพ์ 42
])

# ROC AUC จากการตรวจสอบไขว้แบบ 5-fold (imputer ถูก fit ใหม่ในทุก fold จึงไม่มี data leakage)
auc = cross_val_score(pipe, X, y, cv=5, scoring="roc_auc")          # ช่องที่ 3: พิมพ์ "roc_auc" (รวมเครื่องหมายอัญประกาศ)
print(f"Cross-validated ROC AUC: {auc.mean():.3f} ± {auc.std():.3f}")

# แบ่ง train/test แบบ stratified หนึ่งครั้ง เพื่อคำนวณ sensitivity และ specificity ที่เกณฑ์ตัดสินมาตรฐาน (0.5)
X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
pipe.fit(X_tr, y_tr)                                          # ค่า median คำนวณจากผู้ป่วยในชุดฝึกเท่านั้น
tn, fp, fn, tp = confusion_matrix(y_te, pipe.predict(X_te)).ravel()
sensitivity =tp / (tp + fn)            # ช่องที่ 4: ผู้ป่วยโรคเบาหวานที่ตรวจพบ / ผู้ป่วยโรคเบาหวานทั้งหมด -> พิมพ์  tp / (tp + fn)
specificity =tn / (tn + fp)            # ช่องที่ 5: ผู้ไม่เป็นโรคที่จำแนกถูกต้อง / ผู้ไม่เป็นโรคทั้งหมด -> พิมพ์  tn / (tn + fp)
print(f"ชุดทดสอบ: TP={tp} FN={fn} TN={tn} FP={fp}")
print(f"Sensitivity: {sensitivity:.3f}, Specificity: {specificity:.3f}")

Cross-validated ROC AUC: 0.830 ± 0.035
ชุดทดสอบ: TP=29 FN=25 TN=85 FP=15
Sensitivity: 0.537, Specificity: 0.850


In [ ]:
# @title ✅ เฉลย — หากไม่สามารถทำแบบฝึกหัดได้ ให้รันเซลล์นี้ (ดับเบิลคลิกเพื่อแสดงโค้ด) { display-mode: "form" }
# ไปป์ไลน์อ้างอิง (ไม่มี data leakage, ทำซ้ำได้, มีตัวชี้วัดสำหรับการตรวจคัดกรอง)
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import cross_val_score, train_test_split
from sklearn.metrics import confusion_matrix

ZERO_MEANS_MISSING = ["Glucose", "BloodPressure", "SkinThickness", "Insulin", "BMI"]
X = df.drop(columns=["Outcome"]).replace({c: {0: np.nan} for c in ZERO_MEANS_MISSING})
y = df["Outcome"]

pipe = Pipeline([
    ("imputer", SimpleImputer(strategy="median")),
    ("model", RandomForestClassifier(n_estimators=300, random_state=42)),
])

auc = cross_val_score(pipe, X, y, cv=5, scoring="roc_auc")
print(f"Cross-validated ROC AUC: {auc.mean():.3f} ± {auc.std():.3f}")

X_tr, X_te, y_tr, y_te = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)
pipe.fit(X_tr, y_tr)
tn, fp, fn, tp = confusion_matrix(y_te, pipe.predict(X_te)).ravel()
sensitivity = tp / (tp + fn)
specificity = tn / (tn + fp)
print(f"ชุดทดสอบ: TP={tp} FN={fn} TN={tn} FP={fp}")
print(f"Sensitivity: {sensitivity:.3f}, Specificity: {specificity:.3f}")

💡 **คำอธิบายผลลัพธ์** เมื่อเปรียบเทียบกับรายการตรวจสอบ ไปป์ไลน์อ้างอิงแก้ไขข้อ E (เติมค่าภายใน `Pipeline`) ข้อ F/G (กำหนด `random_state=42` ทุกตำแหน่ง ผลจึงเหมือนเดิมเมื่อรันซ้ำ) ข้อ H (การตรวจสอบไขว้แบบ 5-fold) และข้อ A–C (รายงาน ROC AUC, sensitivity และ specificity) ให้สังเกตค่า sensitivity ที่เกณฑ์ตัดสินมาตรฐาน 0.5 ซึ่งแสดงว่าโมเดลยัง **พลาดผู้ป่วยโรคเบาหวานในสัดส่วนสูง** (FN) ซึ่งไม่เหมาะสมสำหรับการตรวจคัดกรอง ขั้นที่ 4 จะแก้ไขประเด็นนี้

> สำหรับชุดข้อมูลนี้ การรั่วไหลของข้อมูลจากการเติมค่าก่อนแบ่งข้อมูลทำให้คะแนนเปลี่ยนเพียงเล็กน้อย แต่ในข้อมูลอื่น (เช่น การปรับสเกล การคัดเลือกตัวแปร (feature selection) หรือ oversampling ก่อนแบ่งข้อมูล หรือผู้ป่วยรายเดียวกันที่มารับบริการหลายครั้งปรากฏทั้งในชุดฝึกและชุดทดสอบ) การรั่วไหลของข้อมูลอาจทำให้ผลสูงเกินจริงอย่างมาก จึงควรปฏิบัติอย่างถูกต้องจนเป็นนิสัย

## ขั้นที่ 4 — ปรับปรุง: การเลือกเกณฑ์ตัดสินสำหรับการตรวจคัดกรอง

โมเดลให้ค่า *ความน่าจะเป็น (probability)* ของการเป็นโรคเบาหวานสำหรับผู้ป่วยแต่ละราย โดยค่าเริ่มต้น ผู้ที่มีค่าเกิน **0.5** จะถูกจำแนกว่ามีความเสี่ยง สำหรับการตรวจคัดกรอง โดยทั่วไปจะยอมรับผลบวกลวงที่เพิ่มขึ้นเพื่อลดการพลาดผู้ป่วย จึงใช้ **เกณฑ์ตัดสิน (threshold) ที่ต่ำลง**

เซลล์ด้านล่างแสดงค่า sensitivity และ specificity ที่เกณฑ์ตัดสินหลายค่า ซึ่งประเมิน **จากผู้ป่วยในชุดฝึกเท่านั้น** ด้วยการตรวจสอบไขว้ (การเลือกเกณฑ์จากชุดทดสอบถือเป็นการรั่วไหลของข้อมูลอีกรูปแบบหนึ่ง)

**⭐ แบบฝึกหัดที่ 3 (Exercise 3) — แก้ไขตัวเลขเพียงค่าเดียว:** กำหนด `THRESHOLD` เป็นค่าจากตารางที่ให้ **sensitivity ≥ 0.80** และมี specificity สูงที่สุด จากนั้นเซลล์จะประเมินค่าดังกล่าวกับผู้ป่วยในชุดทดสอบหนึ่งครั้ง

*(สำหรับผู้เรียนที่เขียนโปรแกรมได้: สามารถส่งคำสั่งถึงผู้ช่วยปัญญาประดิษฐ์แทน เช่น "Add threshold tuning with cross_val_predict on the training set so sensitivity is at least 0.80, then report test sensitivity and specificity" แล้วตรวจสอบคำตอบด้วยรายการตรวจสอบ)*

In [20]:
# ⭐ Exercise 3 — แบบฝึกหัดที่ 3: เลือกเกณฑ์ตัดสินสำหรับการตรวจคัดกรอง (แก้ไขตัวเลขเพียงค่าเดียว)
from sklearn.model_selection import cross_val_predict

# ความน่าจะเป็นแบบ out-of-fold ของผู้ป่วยในชุดฝึกเท่านั้น (ไม่ใช้ข้อมูลชุดทดสอบ)
p_train_oof = cross_val_predict(pipe, X_tr, y_tr, cv=5, method="predict_proba")[:, 1]
table = pd.DataFrame([
    {"threshold": t,
     "sensitivity": ((p_train_oof >= t) & (y_tr == 1)).sum() / (y_tr == 1).sum(),
     "specificity": ((p_train_oof <  t) & (y_tr == 0)).sum() / (y_tr == 0).sum()}
    for t in [0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5]]).round(3)
print(table.to_string(index=False))

THRESHOLD = 0.3     # TODO: ระบุเกณฑ์จากตารางที่ sensitivity >= 0.80 และ specificity สูงที่สุด (คำใบ้: ประมาณ 0.3)

# ประเมินเกณฑ์ที่เลือกกับผู้ป่วยในชุดทดสอบที่แยกไว้ หนึ่งครั้ง
p_test = pipe.predict_proba(X_te)[:, 1]
flag = p_test >= THRESHOLD
test_sens = (flag & (y_te == 1)).sum() / (y_te == 1).sum()
test_spec = (~flag & (y_te == 0)).sum() / (y_te == 0).sum()
print(f"\nเกณฑ์ {THRESHOLD}: sensitivity ของชุดทดสอบ {test_sens:.3f}, specificity ของชุดทดสอบ {test_spec:.3f}")

 threshold  sensitivity  specificity
      0.20        0.883        0.528
      0.25        0.846        0.598
      0.30        0.808        0.670
      0.35        0.766        0.730
      0.40        0.720        0.782
      0.45        0.654        0.822
      0.50        0.593        0.858

เกณฑ์ 0.3: sensitivity ของชุดทดสอบ 0.778, specificity ของชุดทดสอบ 0.670


In [ ]:
# @title ✅ เฉลย — หากไม่สามารถทำแบบฝึกหัดได้ ให้รันเซลล์นี้ (ดับเบิลคลิกเพื่อแสดงโค้ด) { display-mode: "form" }
# เลือกเกณฑ์โดยอัตโนมัติ: เกณฑ์สูงสุดที่ sensitivity (out-of-fold ของชุดฝึก) >= 0.80
from sklearn.model_selection import cross_val_predict

p_train_oof = cross_val_predict(pipe, X_tr, y_tr, cv=5, method="predict_proba")[:, 1]
table = pd.DataFrame([
    {"threshold": t,
     "sensitivity": ((p_train_oof >= t) & (y_tr == 1)).sum() / (y_tr == 1).sum(),
     "specificity": ((p_train_oof <  t) & (y_tr == 0)).sum() / (y_tr == 0).sum()}
    for t in [0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5]]).round(3)
print(table.to_string(index=False))

ok = table[table["sensitivity"] >= 0.80]
THRESHOLD = float(ok["threshold"].max()) if len(ok) else 0.2   # เกณฑ์สูงสุด = specificity สูงที่สุด

p_test = pipe.predict_proba(X_te)[:, 1]
flag = p_test >= THRESHOLD
test_sens = (flag & (y_te == 1)).sum() / (y_te == 1).sum()
test_spec = (~flag & (y_te == 0)).sum() / (y_te == 0).sum()
print(f"\nเกณฑ์ {THRESHOLD}: sensitivity ของชุดทดสอบ {test_sens:.3f}, specificity ของชุดทดสอบ {test_spec:.3f}")

💡 **คำอธิบายผลลัพธ์** การลดเกณฑ์ตัดสินทำให้ตรวจพบผู้ป่วยโรคเบาหวานได้มากขึ้น (sensitivity เพิ่มขึ้น) แลกกับผลบวกลวงที่เพิ่มขึ้น (specificity ลดลง) โดยค่า ROC AUC ไม่เปลี่ยนแปลง มีเพียงจุดปฏิบัติงาน (operating point) ที่เปลี่ยนไป การเลือกจุดสมดุลนี้เป็น **การตัดสินใจทางคลินิก** (ความเสียหายจากการพลาดผู้ป่วยเทียบกับภาระของการตรวจเพิ่มเติม) มิใช่ประเด็นทางการเขียนโปรแกรม ตัวเลขจากชุดทดสอบมีความแปรปรวน (ชุดทดสอบมีผู้ป่วยโรคเบาหวานประมาณ 54 ราย) จึงอาจแตกต่างจากตารางเพียงเล็กน้อย

## ขั้นที่ 5 — การสะท้อนคิด (ชิ้นงานที่ต้องส่ง)

**Q1.** จงระบุสิ่งที่ผู้ช่วยปัญญาประดิษฐ์ทำ **ผิดพลาดหรือละเว้น** ในครั้งแรก โดยอ้างอิงสัญญาณเตือนที่พบจากรายการตรวจสอบ

✍️ **คำตอบ:**ยังไม่ให้ผลลัพธ์ที่เหมาะสมตั้งแต่ครั้งแรก แม้โค้ดจะรันและให้ค่า accuracy ได้ เพราะพบสัญญาณเตือนหลายอย่าง ได้แก่ โมเดลช่วงแรกมี Sensitivity = 0 หมายความว่าไม่สามารถตรวจพบผู้ป่วยเบาหวานได้เลย ทั้งที่ accuracy ประมาณ 0.65 ซึ่งใกล้เคียงกับสัดส่วนผู้ที่ไม่เป็นเบาหวานในข้อมูลประมาณ 65% นอกจากนี้ยังโค้ดเดิมทำการเติมค่าที่หายไปด้วย median ก่อนแบ่ง train/test ทำให้เกิดความเสี่ยงของ data leakage ดังนั้นจึงไม่ควรพิจารณา accuracy เพียงอย่างเดียวในการประเมินโมเดลสำหรับการคัดกรองโรค

**Q2.** จงเปรียบเทียบตัวเลขที่โค้ดจากปัญญาประดิษฐ์รายงาน (เช่น accuracy, recall ของคลาส 1) กับตัวเลขที่ไปป์ไลน์อ้างอิงรายงาน (ROC AUC จากการตรวจสอบไขว้, sensitivity, specificity) ผู้เรียนพิจารณาว่าตัวเลขชุดใดเหมาะสมสำหรับการตีพิมพ์ เพราะเหตุใด

✍️ **คำตอบ:** ผลจากโมเดลที่ผู้ช่วยปัญญาประดิษฐ์รายงานภายหลังได้ Accuracy = 0.747, Recall/Sensitivity ของคลาส 1 = 0.673, Specificity = 0.788 และ ROC AUC = 0.834 อย่างไรก็ตาม เมื่อตรวจสอบอย่างเหมาะสม ควรใช้ Pipeline และ cross-validation เพื่อป้องกัน data leakage และประเมิน ROC AUC จากหลาย fold แทนการอาศัยการแบ่งข้อมูลเพียงครั้งเดียว
สำหรับงาน screening ตัวชี้วัดที่สำคัญที่สุดคือ Sensitivity เพราะต้องการลดจำนวนผู้ป่วยจริงที่ถูกพลาดหรือ False Negative ให้มากที่สุด ส่วน Specificity ใช้ช่วยประเมินว่ามีผู้ที่ไม่ได้เป็นโรคถูกแจ้งเตือนผิดมากเพียงใด และ ROC AUC ใช้บอกความสามารถโดยรวมของโมเดลในการแยกผู้ป่วยกับผู้ไม่ป่วยในหลาย threshold ดังนั้นในการคัดกรองควรพิจารณา Sensitivity เป็นหลัก ร่วมกับ Specificity และ ROC AUC มากกว่าดู Accuracy เพียงค่าเดียว

**Q3.** จงเขียนคำสั่งในขั้นที่ 1 ใหม่ให้ **ดีขึ้น** เพื่อป้องกันปัญหาที่พบ (เขียนเป็นภาษาไทยหรือภาษาอังกฤษ) *(คำใบ้: ระบุให้ใช้ Pipeline เพื่อให้การเติมค่าคำนวณจากชุดฝึกเท่านั้น กำหนด `random_state` ใช้การตรวจสอบไขว้ และระบุตัวชี้วัดที่ต้องการ ได้แก่ ROC AUC, sensitivity และ specificity)* ไม่บังคับ: ทดลองส่งคำสั่งใหม่ให้ผู้ช่วยปัญญาประดิษฐ์ และบันทึกว่าผ่านรายการตรวจสอบหรือไม่

✍️ **คำตอบ:** สร้างโมเดล Random Forest สำหรับทำนายโรคเบาหวาน โดยให้แทนค่า 0 ที่ไม่สมเหตุสมผลใน Glucose, BloodPressure, SkinThickness, Insulin และ BMI เป็น missing value จากนั้นแบ่งข้อมูลเป็น train/test แบบ stratified ก่อน preprocessing ใช้ Pipeline ที่มี SimpleImputer(strategy="median") และ RandomForestClassifier โดยกำหนด random_state=42 เพื่อให้ผลทำซ้ำได้ ประเมินโมเดลด้วย 5-fold cross-validation และรายงาน ROC AUC รวมทั้งรายงาน sensitivity และ specificity บน test set ห้ามใช้ข้อมูล test set ในการคำนวณค่าที่ใช้เตรียมข้อมูลหรือฝึกโมเดล

**Q4.** โปรดระบุเกณฑ์ตัดสินที่เลือกในแบบฝึกหัดที่ 3 และพิจารณาว่าเหมาะสมสำหรับการใช้ในหน่วยตรวจคัดกรองจริงหรือไม่ ผู้ใดควรเป็นผู้ตัดสินใจเรื่องสมดุลระหว่างการพลาดผู้ป่วยกับผลบวกลวง

✍️ **คำตอบ:** ในแบบฝึกหัดที่ 3 มีการลด threshold จากค่ามาตรฐาน 0.5 ลงมาประมาณ 0.3 เพื่อเพิ่ม Sensitivity ให้ได้อย่างน้อยประมาณ 0.80 การลด threshold ทำให้โมเดลมีแนวโน้มตรวจพบผู้ป่วยเบาหวานมากขึ้น จึงช่วยลด False Negative ซึ่งเป็นสิ่งสำคัญสำหรับหน่วยตรวจคัดกรอง เพราะการพลาดผู้ป่วยจริงอาจทำให้ผู้ป่วยไม่ได้รับการตรวจเพิ่มเติมหรือการดูแลที่เหมาะสม
อย่างไรก็ตาม การลด threshold จะทำให้ False Positive เพิ่มขึ้นและ Specificity ลดลง ดังนั้น threshold ประมาณ 0.3 อาจเหมาะสำหรับงาน screening หากผลจริงแสดงว่า Sensitivity ≥ 0.80 และ Specificity ยังอยู่ในระดับที่หน่วยงานยอมรับได้ แต่ไม่ควรสรุปว่าค่านี้เหมาะสมสำหรับการใช้งานจริงจากข้อมูลชุดนี้เพียงอย่างเดียว ควรมีการตรวจสอบกับข้อมูลภายนอกและพิจารณาผลกระทบทางคลินิกร่วมด้วย

**Q5.** จงอธิบายความเสี่ยงของ vibe coding ในงานวิจัย **ทางคลินิก** หนึ่งประการ ตามความเข้าใจของผู้เรียน (เช่น ประเด็นการรั่วไหลของข้อมูล ความสามารถในการทำซ้ำ หรือความเป็นส่วนตัวของผู้ป่วย)

✍️ **คำตอบ:** การใช้ vibe coding หรือการให้ AI ช่วยสร้างโค้ดอย่างรวดเร็วในงานวิจัยทางการแพทย์มีประโยชน์ในการลดเวลาเขียนโปรแกรม แต่มีความเสี่ยงหากผู้ใช้ไม่ตรวจสอบหลักการของโค้ด ตัวอย่างจากงานนี้คือ AI สามารถสร้างโค้ดที่รันได้และให้ accuracy ที่ดูเหมือนสมเหตุสมผล แต่กลับมี data leakage เพราะทำ imputation ก่อนแบ่ง train/test และในช่วงแรกโมเดลมี Sensitivity = 0 ซึ่งถ้ามองเพียง accuracy อาจทำให้เข้าใจผิดว่าโมเดลมีประสิทธิภาพ
อีกความเสี่ยงคือ reproducibility หากไม่ได้กำหนด random_state ผลอาจเปลี่ยนเมื่อรันซ้ำ รวมถึงผู้ใช้ที่ไม่มีพื้นฐานด้าน coding อาจไม่สามารถตรวจพบข้อผิดพลาดที่ซ่อนอยู่ได้ ดังนั้น AI ควรใช้เป็นเครื่องมือช่วยเขียนโค้ด ไม่ใช่เป็นผู้ตัดสินความถูกต้องของการวิเคราะห์ ผู้ใช้ยังจำเป็นต้องตรวจสอบขั้นตอน preprocessing, data leakage, วิธีแบ่งข้อมูล, metrics และความเหมาะสมทางคลินิกก่อนนำผลไปใช้จริง
สรุปใจความของงานนี้: สิ่งสำคัญไม่ใช่แค่ “โค้ดรันได้หรือไม่” แต่ต้องตรวจสอบด้วยว่า กระบวนการวิเคราะห์ถูกต้อง ไม่มีข้อมูลรั่วไหล ผลทำซ้ำได้ และ metric ที่เลือกสอดคล้องกับวัตถุประสงค์ทางคลินิก โดยเฉพาะงาน screening ที่ควรให้ความสำคัญกับ Sensitivity มากกว่า Accuracy เพียงอย่างเดียว

## 🏆 โจทย์ท้าทายบนตารางอันดับ (Leaderboard) — ไม่บังคับ สำหรับผู้เรียนที่เขียนโปรแกรมได้

**ส่วนนี้ไม่บังคับ และไม่นับเป็นส่วนหนึ่งของชิ้นงานที่ต้องส่ง** เกณฑ์คัดกรองที่ผู้เรียนสร้างใน Lab 1 เป็นค่าอ้างอิง (baseline) แรก ในส่วนนี้ให้ส่ง **ความน่าจะเป็น (probability)** จากไปป์ไลน์ที่ *ไม่มีการรั่วไหลของข้อมูล* สำหรับ **ผู้ป่วย 154 รายชุดเดิมที่แยกไว้** **ตัวชี้วัด: ROC AUC** (ตารางอันดับแสดง sensitivity/specificity ที่เกณฑ์ 0.5 ประกอบด้วย) ไฟล์ชื่อ `lab02_submission.csv` มีส่วนหัว `id,prob` (prob = ความน่าจะเป็นของการเป็นโรคเบาหวาน) ให้ fit โมเดลกับแถว *อื่น* เท่านั้น

การส่งผล ให้ระบุ `NAME` และ `JOIN_CODE` (ผู้สอนประกาศในการเรียนการสอนผ่าน Zoom) ในเซลล์ ⚙️ การตั้งค่า แล้วรันเซลล์นั้นอีกครั้ง เซลล์ค่าอ้างอิงด้านล่างสามารถใช้งานได้ทันทีด้วยไปป์ไลน์อ้างอิงจากขั้นที่ 3

> **หลักจริยธรรมทางวิชาการ (Honour code)** label ที่ซ่อนไว้มาจากชุดข้อมูลสาธารณะ การค้นหาคำตอบจึงทำได้ง่ายแต่ไม่เกิดประโยชน์ ตารางอันดับมีไว้เพื่อสนับสนุนการเรียนรู้ และโน้ตบุ๊กของผู้เรียนจะได้รับการประเมินด้วย ห้ามแก้ไขค่าผลการทำนายด้วยตนเอง

In [ ]:
# 🏆 ส่งผลค่าอ้างอิง (baseline) — ไปป์ไลน์อ้างอิง fit กับแถวที่ไม่ใช่ชุดทดสอบเท่านั้น
from sklearn.base import clone

test_ids = pd.read_csv(lab_data("pima/test_ids.csv"))["id"]
train_mask = ~X.index.isin(test_ids)

baseline_pipe = clone(pipe)                                   # สำเนาไปป์ไลน์อ้างอิงจากขั้นที่ 3 (ยังไม่ผ่านการฝึก)
baseline_pipe.fit(X[train_mask], y[train_mask])               # ห้าม fit กับแถวของชุดทดสอบ
prob = baseline_pipe.predict_proba(X.loc[test_ids])[:, 1]

pd.DataFrame({"id": test_ids.values, "prob": prob}).to_csv("lab02_submission.csv", index=False)
print("บันทึกไฟล์ lab02_submission.csv เรียบร้อย จำนวน", len(prob), "แถว")
if NAME and JOIN_CODE:
    submit_to_leaderboard("lab02_submission.csv")
else:
    print("หากต้องการส่งผล โปรดระบุ NAME และ JOIN_CODE ในเซลล์การตั้งค่า (ไม่บังคับ)")

**แนวทางการทำคะแนนให้สูงกว่าค่าอ้างอิง** (สามารถปรึกษาผู้ช่วยปัญญาประดิษฐ์ และต้องตรวจสอบทุกข้อเสนอด้วยรายการตรวจสอบก่อนนำไปใช้)
- เปรียบเทียบโมเดลด้วยการตรวจสอบไขว้แบบ 5-fold *ชุดเดียวกัน* บนแถวของชุดฝึก: `LogisticRegression` (ใส่ `StandardScaler` **ภายใน** Pipeline), `HistGradientBoostingClassifier` (รองรับค่า NaN โดยตรง) หรือ random forest ที่ปรับค่า (`min_samples_leaf`, `max_features`)
- เพิ่มตัวแปรบ่งชี้ค่าที่ขาดหาย (`SimpleImputer(add_indicator=True)`) เนื่องจากการ "ไม่ได้ตรวจวัด insulin" อาจเป็นข้อมูลที่มีความหมาย
- สร้างตัวแปรใหม่ (feature engineering) ภายในไปป์ไลน์ (เช่น glucose × BMI) หรือเฉลี่ยความน่าจะเป็นจากโมเดล 2–3 โมเดล
- ปรับไฮเปอร์พารามิเตอร์ด้วย `GridSearchCV` บนแถวของชุดฝึกเท่านั้น ทั้งนี้ การปรับเกณฑ์ตัดสินไม่ทำให้ค่า ROC AUC เปลี่ยนแปลง

In [ ]:
# 🏆 ไม่บังคับ — ไปป์ไลน์ที่ผู้เรียนปรับปรุง (ต้องไม่มี data leakage; หากกำหนดเป็น None จะใช้ค่าอ้างอิงต่อไป)
my_pipe = None      # ตัวอย่าง: Pipeline([("imputer", SimpleImputer(strategy="median", add_indicator=True)),
                    #                     ("model", HistGradientBoostingClassifier(random_state=42))])

if my_pipe is not None:
    cv_auc = cross_val_score(my_pipe, X[train_mask], y[train_mask], cv=5, scoring="roc_auc")
    print(f"ไปป์ไลน์ของผู้เรียน — CV ROC AUC บนแถวของชุดฝึก: {cv_auc.mean():.3f} ± {cv_auc.std():.3f}")
    my_pipe.fit(X[train_mask], y[train_mask])                 # ห้าม fit กับแถวของชุดทดสอบ
    prob = my_pipe.predict_proba(X.loc[test_ids])[:, 1]
    pd.DataFrame({"id": test_ids.values, "prob": prob}).to_csv("lab02_submission.csv", index=False)
    print("บันทึกไฟล์ lab02_submission.csv เรียบร้อย")
    if NAME and JOIN_CODE:
        submit_to_leaderboard("lab02_submission.csv")
else:
    print("ยังไม่ได้กำหนดไปป์ไลน์ที่ปรับปรุง — ใช้ไฟล์ผลการทำนายของค่าอ้างอิงต่อไป")

## ชิ้นงานที่ต้องส่ง (งานเดี่ยว)

1. โน้ตบุ๊กนี้ที่ **รันครบทุกเซลล์ตามลำดับจากบนลงล่าง** (เมนู Runtime → Run all ต้องทำงานได้โดยไม่เกิดข้อผิดพลาด โดยใช้เซลล์ ✅ เฉลย ตามความจำเป็น)
2. คำตอบในเซลล์ ✍️ ได้แก่ รายการตรวจสอบ (ขั้นที่ 2) และคำถาม Q1–Q5 (ขั้นที่ 5) พร้อมบันทึกคำสั่งที่ใช้ไว้ในเซลล์แบบฝึกหัดที่ 1

ให้ส่งงานตามช่องทางที่รายวิชากำหนด *ไม่บังคับ:* ผู้เรียนที่เข้าร่วมโจทย์ท้าทาย 🏆 สามารถระบุชื่อบนตารางอันดับและคะแนนสูงสุดไว้ในส่วนนี้ โดยไม่นับเป็นข้อกำหนดของชิ้นงาน